e2e/cluster/cluster.go

```diff
@@ -0,0 +1,503 @@
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
+)
+
+// WARNING: if you modify cluster configuration, please change the version below
+// this will avoid potential conflicts with tests running on other branches
+// there is no strict rules or a hidden meaning for the version
+// testClusterNamePrefix is also used for versioning cluster configurations
+const testClusterNamePrefix = "abe2e-v20240704-"
+
+var (
+	clusterKubenet       *Cluster
+	clusterKubenetAirgap *Cluster
+	clusterAzureNetwork  *Cluster
+
+	clusterKubenetError       error
+	clusterKubenetAirgapError error
+	clusterAzureNetworkError  error
+
+	clusterKubenetOnce       sync.Once
+	clusterKubenetAirgapOnce sync.Once
+	clusterAzureNetworkOnce  sync.Once
+)
+
+type Cluster struct {
+	Model *armcontainerservice.ManagedCluster
+	Kube  *Kubeclient
+}
+
+type Kubeclient struct {
+	Dynamic client.Client
+	Typed   kubernetes.Interface
+	Rest    *rest.Config
+}
+
+// Returns true if the cluster is configured with Azure CNI
+func (c *Cluster) IsAzureCNI() (bool, error) {
+	if c.Model.Properties.NetworkProfile != nil {
+		return *c.Model.Properties.NetworkProfile.NetworkPlugin == armcontainerservice.NetworkPluginAzure, nil
+	}
+	return false, fmt.Errorf("cluster network profile was nil: %+v", c.Model)
+}
+
+// Returns the maximum number of pods per node of the cluster's agentpool
+func (c *Cluster) MaxPodsPerNode() (int, error) {
+	if len(c.Model.Properties.AgentPoolProfiles) > 0 {
+		return int(*c.Model.Properties.AgentPoolProfiles[0].MaxPods), nil
+	}
+	return 0, fmt.Errorf("cluster agentpool profiles were nil or empty: %+v", c.Model)
+}
+
+// Same cluster can be attempted to be created concurrently by different tests
+// sync.Once is used to ensure that only one cluster for the set of tests is created
+func ClusterKubenet(ctx context.Context) (*Cluster, error) {
+	clusterKubenetOnce.Do(func() {
+		clusterKubenet, clusterKubenetError = createNewClusterWithClient(ctx, getKubenetClusterModel(testClusterNamePrefix+"kubenet-v1"))
+	})
+	return clusterKubenet, clusterKubenetError
+}
+
+func ClusterKubenetAirgap(ctx context.Context) (*Cluster, error) {
+	clusterKubenetAirgapOnce.Do(func() {
+		cluster, err := createNewClusterWithClient(ctx, getKubenetClusterModel(testClusterNamePrefix+"kubenet-airgap"))
+		if err == nil {
+			err = addAirgapNetworkSettings(ctx, cluster)
+		}
+		clusterKubenetAirgap, clusterKubenetAirgapError = cluster, err
+	})
+	return clusterKubenetAirgap, clusterKubenetAirgapError
+}
+
+func ClusterAzureNetwork(ctx context.Context) (*Cluster, error) {
+	clusterAzureNetworkOnce.Do(func() {
+		clusterAzureNetwork, clusterAzureNetworkError = createNewClusterWithClient(ctx, getAzureNetworkClusterModel(testClusterNamePrefix+"azure-network"))
+	})
+	return clusterAzureNetwork, clusterAzureNetworkError
+}
```

Comment:
New code is mostly in this file.

Some of the existing functions also has been moved here. So, the airgap creation logic can be referenced in scenarios without causing circular dependency issues.